app/lib/admin/backend.dart

```diff
@@ -0,0 +1,206 @@
+import '../account/models.dart';
+import '../package/models.dart';
+import '../publisher/models.dart';
+import '../shared/configuration.dart';
+import '../shared/exceptions.dart';
+
+final _logger = Logger('pub.admin.backend');
+
+/// Sets the admin backend service.
+void registerAdminBackend(AdminBackend backend) =>
+    ss.register(#_adminBackend, backend);
+
+/// The active admin backend service.
+AdminBackend get adminBackend => ss.lookup(#_adminBackend) as AdminBackend;
+
+/// Represents the backend for the admin handling and authentication.
+class AdminBackend {
+  final DatastoreDB _db;
+  AdminBackend(this._db);
+
+  Future<R> _withAdmin<R>(Future<R> fn(AuthenticatedUser user)) async {
+    return await withAuthenticatedUser((u) async {
+      // TODO: remove once [withAuthenticatedUser] is using [User]
+      final user = await accountBackend.lookupUserById(u.userId);
+      if (!activeConfiguration.admins.contains(user.oauthUserId)) {
+        throw AuthorizationException.userIsNotAdminForPubSite();
+      }
+      return await fn(u);
+    });
+  }
+
+  /// List users.
+  Future<api.AdminListUsersResponse> listUsers(
+      {String continuationToken, int limit = 1000}) async {
+    return await _withAdmin((user) async {
+      final query = _db.query<User>()
+        ..order('__key__')
+        ..limit(limit);
+
+      if (continuationToken != null) {
+        Map<String, dynamic> map;
+        try {
+          map = _decodeContinuation(user.userId, continuationToken);
+        } on FormatException catch (_) {
+          throw InvalidInputException.continuationParseError();
+        }
+        InvalidInputException.checkNotNull(map, 'continuationToken');
+
+        final id = map['id'] as String;
+        // NOTE: we should fix https://github.com/dart-lang/gcloud/issues/23
+        //       and remove the toDatastoreKey conversion here.
+        final key =
+            _db.modelDB.toDatastoreKey(_db.emptyKey.append(User, id: id));
+        if (id != null) {
+          query.filter('__key__ >', key);
+        }
+      }
+
+      try {
+        final users = await query.run().toList();
+        return api.AdminListUsersResponse(
+          users: users
+              .map(
+                (u) => api.AdminUserEntry(
+                  userId: u.userId,
+                  email: u.email,
+                  oauthUserId: u.oauthUserId,
+                ),
+              )
+              .toList(),
+          continuationToken: users.length < limit
+              ? null
+              : _encodeContinuation(user.userId, {'id': users.last.userId}),
+        );
+      } catch (e, st) {
+        print(e);
+        print(st);
+        rethrow;
+      }
+    });
+  }
+
+  /// Removes user from the Datastore and updates the packages and other
+  /// entities they may have controlled.
+  Future removeUser(String userId) async {
+    await _withAdmin((u) async {
+      // TODO: remove once [withAuthenticatedUser] is using [User]
```

Comment:
I don't follow -- what is the todo?

Reply:
I'm planning to refactor the `withAuthenticatedUser` callback to return with an `User` object instead of the `AuthenticatedUser` object. If that happens, the line that follows which uses the `AuthenticatedUser.userId` to lookup the `User` won't be needed anymore.

Reply:
You do realize that the user being removed is different from the user making the call, right?

Reply:
Oops :)